File: leetcode/python/000621.task-scheduler.py
import heapq
from collections import Counter, deque


class Solution:
    # Time complexity: O(n*log(k))
    # Each task gets counted once, and gets processed in heap once
    # Space complexity: O(k), where k is number of unique tasks
    # We never store more than k items in any of the data structures
    def leastInterval(self, tasks: list[str], n: int) -> int:
        count = Counter(tasks)
        maxHeap = [-c for c in count.values()]
        heapq.heapify(maxHeap)
        queue = deque()  # (clock cycle when op can be executed again, op count)
        clock = 0

        while maxHeap or queue:
            clock += 1

            if maxHeap:
                c = heapq.heappop(maxHeap)  # O(log(26))
                c += 1  # reduce count
                if c < 0:
                    queue.append((clock + n, c))

            if queue and queue[0][0] == clock:
                _, c = queue.popleft()
                heapq.heappush(maxHeap, c)  # O(log(26))

        return clock

File: leetcode/javascript/000621.task-scheduler.js
/**
 * @param {character[]} tasks
 * @param {number} n
 * @return {number}
 */
var leastInterval = function(tasks, n) {
    const countMap = new Map();
    tasks.forEach((task) => {
        countMap.set(task, (countMap.get(task) ?? 0) + 1)
    })

    let clock = 0;
    const q = new Queue();
    const pq = PriorityQueue.fromArray(Array.from(countMap.values()), (a, b) => b - a);

    while (!q.isEmpty() || !pq.isEmpty()) {
        clock++;

        if (!pq.isEmpty()) {
            c = pq.dequeue();
            c--;
            if (c > 0) {
                q.enqueue([clock + n, c]);
            }
        }

        if (!q.isEmpty() && q.front()[0] === clock) {
            c = q.dequeue()[1];
            pq.enqueue(c);
        }
    }

    return clock;
};
